File: backend/src/main/java/com/englishlearning/backend/service/ListeningLessonService.java
package com.englishlearning.backend.service;
import com.englishlearning.backend.dto.request.ListeningLessonCreateRequest;
import com.englishlearning.backend.dto.request.UpdateListeningLessonRequest;
import com.englishlearning.backend.dto.response.ListeningLessonResponse;
import com.englishlearning.backend.enums.ListeningLessonStatus;

import java.util.List;

public interface ListeningLessonService {
    ListeningLessonResponse create(
            Long teacherId,
            ListeningLessonCreateRequest request
    );
    ListeningLessonResponse update(
            Long teacherId,
            Long lessonId,
            UpdateListeningLessonRequest request
    );
    List<ListeningLessonResponse> getMyLessons(
            Long teacherId
    );
    List<ListeningLessonResponse> getMyLessonsByTopic(Long teacherId, Long topicId);
    ListeningLessonResponse submitForReview(
            Long teacherId,
            Long lessonId
    );
    ListeningLessonResponse reject(
            Long adminId,
            Long lessonId
    );
    List<ListeningLessonResponse> getAll();
    ListeningLessonResponse approve(
            Long adminId,
            Long lessonId
    );
    ListeningLessonResponse publish(
            Long adminId,
            Long lessonId
    );
    ListeningLessonResponse getById(
            Long lessonId
    );
    List<ListeningLessonResponse> getByTopic(
            Long topicId
    );
    List<ListeningLessonResponse> getPublishedByTopic(
            Long topicId
    );

    void hardDelete(Long teacherId, Long lessonId);
    void softDelete(Long adminId, Long lessonId);
    void restore(Long adminId, Long lessonId);
    List<ListeningLessonResponse> getDeletedLessons();
    List<ListeningLessonResponse> getByTopicForAdmin(Long topicId);
    List<ListeningLessonResponse> getAll(ListeningLessonStatus status);
}

File: frontend/src/services/listeningLessonService.js
import axiosClient from "../api/axiosClient";

const listeningLessonService = {
  // ===== TEACHER =====
  create(data) {
    return axiosClient.post("/v1/listening-lessons", data);
  },

  update(id, data) {
    return axiosClient.put(`/v1/listening-lessons/${id}`, data);
  },

  getMyLessons() {
    return axiosClient.get("/v1/listening-lessons/my");
  },

  getMyLessonsByTopic(topicId) {
    return axiosClient.get(`/v1/listening-lessons/topic/${topicId}/my`);
  },

  submit(id) {
    return axiosClient.post(`/v1/listening-lessons/${id}/submit`);
  },

  // ===== TEACHER - HARD DELETE =====
  hardDelete(id) {
    return axiosClient.delete(`/v1/listening-lessons/${id}/hard`);
  },

  // ===== PUBLIC =====
  getById(id) {
    return axiosClient.get(`/v1/listening-lessons/${id}`);
  },

  getByTopic(topicId) {
    return axiosClient.get(`/v1/listening-lessons/topic/${topicId}`);
  },

  getPublishedByTopic(topicId) {
    return axiosClient.get(`/v1/listening-lessons/topic/${topicId}/published`);
  },

  // ===== ADMIN =====
  getAllForAdmin() {
    return axiosClient.get("/v1/listening-lessons/admin");
  },

  // ===== ADMIN - GET BY TOPIC (bao gồm cả bài đã ẩn) =====
  getByTopicForAdmin(topicId) {
    return axiosClient.get(`/v1/listening-lessons/admin/topic/${topicId}`);
  },

  approve(id) {
    return axiosClient.post(`/v1/listening-lessons/admin/${id}/approve`);
  },

  reject(id) {
    return axiosClient.post(`/v1/listening-lessons/admin/${id}/reject`);
  },

  publish(id) {
    return axiosClient.post(`/v1/listening-lessons/admin/${id}/publish`);
  },

  // ===== ADMIN - SOFT DELETE =====
  softDelete(id) {
    return axiosClient.delete(`/v1/listening-lessons/admin/${id}/soft`);
  },

  // ===== ADMIN - RESTORE =====
  restore(id) {
    return axiosClient.post(`/v1/listening-lessons/admin/${id}/restore`);
  },

  // ===== ADMIN - GET DELETED LESSONS =====
  getDeletedLessons() {
    return axiosClient.get("/v1/listening-lessons/admin/deleted");
  },
};

export default listeningLessonService;
